README.md
# CloudFlare Originator
Cloudflare Origin Customizer, allowing to map Cloudflare CDN to Origin paths

<a href="https://www.cloudflare.com/apps/cloudflare-originator/install?source=button">
  <img
    src="https://install.cloudflareapps.com/install-button.png"
    alt="Install Cloudflare Originator with Cloudflare"
    border="0"
    width="150">
</a>

## Overview

This Cloudflare App attempts to replicate AWS CloudFront's convenient capability of mapping path of incoming requests to a different origin path.

Can be used as a standalone worker, see [`workers/worker.js`](./workers/worker.js) for details.

## Features

#### Cloudflare DNS

Adds DNS record used as your Cloudflare-backed CDN.

#### Cloudflare Workers

Adds [Cloudflare Worker](https://developers.cloudflare.com/workers/), which maps all incoming requests matching your Cloudflare CDN hostname to configured Origin path.

> WARNING: When using S3 as Origin, make sure to have S3 URLs publicly reachable, or have appropriate bucket policy. See [S3_as_Origin.md](./S3_as_Origin.md) for example policy.

## Setup

- Fork and clone the repo.
- Make desired changes.
- Install the dependencies with `yarn install` then build the project with `yarn build` (or `npm run build`).
- Next, navigate to [Cloudflare App Creator](https://www.cloudflare.com/apps/developer/app-creator) and upload your project directory.

The App Creator will update automatically on file changes (Chrome only). Once you're done testing, press Create App to submit your app for moderation. Refer to our [Terms of Use](https://www.cloudflare.com/apps/developer/docs/resources/terms-of-use) for more information.

## Usage

- `yarn start` (or `npm start`) Sets up your dev environment and runs Webpack in watch mode.
- `yarn build` (or `npm run build`) Lints your project and compiles your JavaScript and CSS once for release.

### Other Interesting Scripts

- `dev:setup` Add other initialization scripts to your development pipeline here.

The project uses [yarn-run-all](https://www.npmjs.com/package/yarn-run-all) which makes it easy to run tasks in series or parallel using `yarn` or `npm`.

## Details

#### `install.json`

This is where all the [installer options](https://www.cloudflare.com/apps/developer/docs/install-json) are added for the app.

DNS field is used to to configure [Cloudflare DNS Records](https://api.cloudflare.com/#dns-records-for-a-zone-properties). If you don't want to configure DNS, just delete this field.

#### `src/index.js`

This is where the magic happens. Your app starts here.

#### `src/styles.css`

Write your app styles here.

#### `workers/worker.js`

This file is used to add a [Cloudflare Worker](https://developers.cloudflare.com/workers/) to the app. Each app may contain one worker. If you don't want to use workers, just delete this file along with the workers configuration in `install.json`.

#### `media/**`

Contains icons, tile images, and screenshots used in your Cloudflare Apps page.

[Download <code class="inline">media-templates.sketch</code>](https://github.com/CloudflareApps/MediaTemplates/raw/master/media-templates.sketch)

#### `webpack.config.js`

Simple Webpack 4 config using Babel and CSS Loader. Please refrain from modifying the config to minify your built code.

### Troubleshooting

The Cloudflare [developer documentation](https://www.cloudflare.com/apps/developer/docs/getting-started) for examples and API usage.


## Authors

- Anastas Dancha (aka [@anapsix](https://github.com/anapsix)) - original implementation
- Nicole Corbin (aka [@ncorbin](https://github.com/ncorbin)) - logo artwork



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.2] - 2019-05-02
### Added
- getWorkerIP option, to disable checking worker IP, even when debug is on
- cdnCatchAll option, to control request pass-through when path match is
  not found in pathMap, helpful when CDN hostname is same as main site

## [1.0.1] - 2019-04-11
### Added
- return 404 if origin response code is not 200
- use passThroughOnException if worker fails

## [1.0.0] - 2019-04-10
### Added
- Initial revision.
- May contain bugs.
- App creation driven by internal usage/need.
- Your mileage may vary.


src/index.js
// this import statement tells webpack to include styles.css in the build
import css from "./styles.css"

function init() {
  if (!window.addEventListener) return // Check for IE9+

  let options = INSTALL_OPTIONS
  let element

  const message =
    "<h1>CloudFlare Originator.</h1><p>This Cloudflare App attempts to replicates AWS CloudFront's convenient capability of mapping path of incoming requests to a different origin path.</p>"

  const location = {
    selector: "body",
    method: "prepend",
  }

  // updateElement runs every time the options are updated.
  // Most of your code will end up inside this function.
  function updateElement() {
    element = INSTALL.createElement(location, element)

    // Set the app attribute to your app's dash-delimited alias.
    element.setAttribute("app", "origin-path-customizer")
    element.innerHTML = message
  }

  // INSTALL_SCOPE is an object that is used to handle option changes without refreshing the page.
  window.INSTALL_SCOPE = {
    setOptions(nextOptions) {
      options = nextOptions
      updateElement()
    },
  }

  // This code ensures that the app doesn't run before the page is loaded.
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", updateElement)
  } else {
    updateElement()
  }
}

init()


workers/worker.js
/**
 * This worker script allows mapping CDN to Origin arbitrary paths
 */

/**
 * INSTALL_OPTIONS must be defined
 * used by standalone worker
 * For example:
 *  const INSTALL_OPTIONS = {
 *    "debug": false,
 *    "getWorkerIP": false
 *    "cdnCatchAll": true,
 *    "originUrl": "https://myorigin.com",
 *    "cdnDomain": "myzonedomain.com",
 *    "cdnSubDomain": "cdn",
 *    "pathMap": [
 *      { "script/": "static_assets/script/" },
 *      { "style/": "static/style/"}
 *    ]
 *  }
 */

/* SETUP BEGINS */

/**
 * Example 1
 * CDN Path: https://cdn.myzonedomain.com/script/main.js
 * S3 Path: https://myorigin.com/grunt_build/script/main.js
 *
 * Example 2
 * CDN Path: https://cdn.myzonedomain.com/styles/style.css
 * S3 Path: https://myorigin.com/grunt_build/styles/style.css
 */

/* SETUP ENDS */

const VERSION = "1.0.2"
const WORKER_SHORT_NAME = "originator"
const DEBUG = INSTALL_OPTIONS.debug
const GET_WORKER_IP = INSTALL_OPTIONS.getWorkerIP
const CDN_CATCH_ALL = INSTALL_OPTIONS.cdnCatchAll
const originUrl = INSTALL_OPTIONS.originUrl.trim()
const originHost = originUrl.replace(new RegExp("^https?://"), "")
// eslint-disable-next-line
const cndPath = `^https?://${INSTALL_OPTIONS.cdnSubDomain}.${INSTALL_OPTIONS.cdnDomain}/`
const cdnPathRegExp = new RegExp(cndPath)

/* eslint-disable no-param-reassign */

/**
 * used by CloudFlare App Worker
 */
const cdnPathMap = INSTALL_OPTIONS.pathMap.reduce((map, obj) => {
  map[obj.key] = obj.val
  return map
}, {})

/**
 * used by standalone worker
 */
// const cdnPathMap = INSTALL_OPTIONS.pathMap.reduce((map, obj) => {
//   var obj2 = Object.entries(obj).map(([key, value]) => ({key,value}));
//   map[obj2[0].key] = obj2[0].value
//   return map
// }, {})
/* eslint-enable no-param-reassign */

/**
 * Print debug message if DEBUG is enabled
 * @param {string} msg
 */
async function debug(msg) {
  if (DEBUG) {
    if (typeof msg === "string") {
      console.log(`${WORKER_SHORT_NAME} ${VERSION}: ${msg}`)
    } else {
      console.log(msg)
    }
  }
}

debug("== WORKER REQUEST DEBUG ENABLED ==")
debug("CDN Path Map")
debug(cdnPathMap)

/**
 * get worker IP address
 */
async function getWorkerIP() {
  if (DEBUG && GET_WORKER_IP) {
    const ipReq = new Request("http://icanhazip.com")
    const ipRes = await fetch(ipReq)
    return ipRes.text()
  }
  return "0.0.0.0"
}

/**
 * Get CDN to Origin mapped path
 * @params {string} path
 */
function getOriginPath(path) {
  /* eslint-disable no-restricted-syntax */
  for (const key of Object.keys(cdnPathMap)) {
    const cdnPR = new RegExp(`^${key}`)
    if (cdnPR.test(path)) {
      const pathRem = path.replace(cdnPR, "")
      return `${cdnPathMap[key]}${pathRem}`
    }
  }
  /* eslint-enable no-restricted-syntax */
  if (CDN_CATCH_ALL) {
    debug(
      "Cannot find specific path mapping, using " /
        " as prefix, since CDN_CATCH_ALL enabled..",
    )
    return path
  }
  debug(
    "Cannot find specific path mapping, bypassing CDN since CDN_CATCH_ALL is disabled..",
  )
  return false
}

/**
 * Fetch and log a request
 * @param {Request} request
 */
async function fetchAndLog(request) {
  debug("Got request", request)
  const response = await fetch(request)
  debug("Got response", response)
  return response
}

/**
 * Fetch Origin asset
 * @params {string} url
 */
async function getFromOrigin(url) {
  debug(`Fetching from Origin: ${url}`)
  const urlReq = new Request(url)
  urlReq.headers.set("HOST", originHost)
  debug("Origin Request Headers")
  debug(new Map(urlReq.headers))
  const [urlRes, workerIp] = await Promise.all([fetch(urlReq), getWorkerIP()])
  debug(`Worker IP: ${workerIp}`)
  debug(`Origin Response Code: ${urlRes.status}`)
  if (urlRes.status !== 200) {
    debug("Returning 404")
    return new Response("This page is not found.", {
      status: 404,
      statusText: "NotFound",
    })
  }
  return urlRes
}

/* eslint-disable no-restricted-globals */
/**
 * Register Event Listener
 */
addEventListener("fetch", event => {
  event.passThroughOnException()
  if (cdnPathRegExp.test(event.request.url)) {
    debug("CDN request candidate found")
    const requestPath = event.request.url.replace(cdnPathRegExp, "")
    const originPath = getOriginPath(requestPath)
    debug(`CDN to Origin Mapped Path: "${requestPath}" => "${originPath}"`)
    if (originPath !== false) {
      const originResourceUrl = `${originUrl}/${originPath}`
      debug(`Origin URL: ${originResourceUrl}`)
      event.respondWith(getFromOrigin(originResourceUrl))
    }
  } else {
    debug("Default request matched")
    // event.respondWith(fetchAndLog(event.request))
  }
})
/* eslint-enable no-restricted-globals */
